### ISSUE TITLE
“alwaysCreatePr" flag doesn't seem to be working correctly

### ISSUE THREAD

**Issue by sgabler** [shayki5/azure-devops-create-pr-task#126]
After https://github.com/shayki5/azure-devops-create-pr-task/pull/124 was merged, the feature introduced in #118 unfortunately stopped working.

I already set `alwaysCreatePr: false` in my pipeline, but still an empty PR was created with 0 file changes. (The branch was 1 commit ahead, but no actual file changes are there)

Am I using this incorrectly somehow, or is there maybe a bug somewhere? Happy to help debugging, but at first glance I couldn't spot the issue

FYI @Bene2205 @shayki5

**Comment by shayki5** [comment]
Hi @sgabler, this feature merged to master but didn't publish publicly yet. 
Which task version you see in the logs?
And, in the logs what is the output? **0 new commits! perform a Pull Request...**?

**Comment by sgabler** [comment]
sorry about not providing it from the beginning:

```
    Starting: 🤖 Merge release into develop
    ==============================================================================
    Task         : Create Pull Request
    Description  : Automatically create a Pull Request for Azure DevOps or GitHub repository from Build or Release pipeline, supports also multi target branch.
    Version      : 1.2.272
    Author       : Shayki Abramczyk
    Help         : [More Information](https://github.com/shayki5/azure-devops-create-pr-task/blob/master/README.md)
    ==============================================================================
    The repository is: <removed>
    The Source Branch is: refs/heads/release
    The Target Branch is: refs/heads/develop
    The Title is: 🤖 Merge release into develop
    The Description is: Automatically created by "Azure DevOps Create Pull Request Task"
    Is Reviewers are: 
    Is Draft Pull Request: False
    Auto-Complete: True
    Link Work Items: True
    Bypass: False
    Bypass Reason: 
    DeleteSourceBranch ist set to: False
    1 new commits! perform a Pull Request...
    {
        "forkSource":  "",
        "sourceRefName":  "refs/heads/release",
        "title":  "🤖 Merge release into develop",
        "WorkItemRefs":  [
                             null
                         ],
        "description":  "Automatically created by \"Azure DevOps Create Pull Request Task\"",
        "reviewers":  "",
        "isDraft":  "False",
        "targetRefName":  "refs/heads/develop"
    }
    *************************
    ******** Success ********
    *************************
    Pull Request 3220 created.
    ##[warning]Can't set Auto Complete to PR 3220.
    ##[warning]{"$id":"1","innerException":null,"message":"TF401027: You need the Git 'GenericContribute' permission to perform this action. Details: identity 'Build\\11400df4-d5a0-4a8a-b46c-a4d5632b067a', scope 'branch'.","typeName":"Microsoft.TeamFoundation.Git.Server.GitNeedsPermissionException, Microsoft.TeamFoundation.Git.Server","typeKey":"GitNeedsPermissionException","errorCode":0,"eventId":3000}
    ##[warning]The remote server returned an error: (403) Forbidden.
    Finishing: 🤖 Merge release into develop
```

So this run used 1.2.272. Previously it worked, but last time I executed this task it used version 1.2.264. (It was paused due to some other issues, so today I tested it again after some weeks.)

**Comment by sgabler** [comment]
The 1 commit mentioned above looks like this:

Merged PR 3201: Merge develop into release

But the file diff is empty

**Comment by shayki5** [comment]
Ok, this is another issue, not related to the merged PR...
Can you run the pipeline with `system.debug` = `true` and share the logs?

**Comment by sgabler** [comment]
Sorry, maybe I jumped to conclusions :sweat_smile:

```
    ##[debug]Evaluating condition for step: '🤖 Merge release into develop'
    ##[debug]Evaluating: SucceededNode()
    ##[debug]Evaluating SucceededNode:
    ##[debug]=> True
    ##[debug]Result: True
    Starting: 🤖 Merge release into develop
    ==============================================================================
    Task         : Create Pull Request
    Description  : Automatically create a Pull Request for Azure DevOps or GitHub repository from Build or Release pipeline, supports also multi target branch.
    Version      : 1.2.277
    Author       : Shayki Abramczyk
    Help         : [More Information](https://github.com/shayki5/azure-devops-create-pr-task/blob/master/README.md)
    ==============================================================================
    ##[debug]VstsTaskSdk 0.10.0 commit 787a46ec0a2df5b4d12c2e801bd3f319975c054c
    ##[debug]Entering RunTask.
    ##[debug]INPUT_SOURCEBRANCH: 'release'
    ##[debug]INPUT_TARGETBRANCH: 'develop'
    ##[debug]INPUT_TITLE: '🤖 Merge release into develop'
    ##[debug]INPUT_DESCRIPTION: 'Automatically created by "Azure DevOps Create Pull Request Task"'
    ##[debug]INPUT_REVIEWERS (empty)
    ##[debug]INPUT_REPOTYPE: 'Azure DevOps'
    ##[debug]INPUT_ISDRAFT: 'false'
    ##[debug] Converted to bool: False
    ##[debug]INPUT_AUTOCOMPLETE: 'true'
    ##[debug] Converted to bool: True
    ##[debug]INPUT_MERGESTRATEGY: 'Merge (No fast-forward)'
    ##[debug]INPUT_DELETESOURCH (empty)
    ##[debug] Converted to bool: False
    ##[debug]INPUT_DELETESOURCE: 'false'
    ##[debug] Converted to bool: False
    ##[debug]INPUT_COMMITMESSAGE (empty)
    ##[debug]INPUT_TRANSITIONWORKITEMS: 'false'
    ##[debug] Converted to bool: False
    ##[debug]INPUT_LINKWORKITEMS: 'true'
    ##[debug] Converted to bool: True
    ##[debug]INPUT_PROJECTID: '<removed>'
    ##[debug]INPUT_GITREPOSITORYID: '<removed>'
    ##[debug]INPUT_GITHUBREPOSITORY: '<removed>'
    ##[debug]INPUT_PASSPULLREQUESTIDBACKTOADO: 'false'
    ##[debug] Converted to bool: False
    ##[debug]INPUT_ISFORKED: 'false'
    ##[debug] Converted to bool: False
    ##[debug]INPUT_BYPASSPOLICY: 'false'
    ##[debug] Converted to bool: False
    ##[debug]INPUT_BYPASSREASON (empty)
    ##[debug]INPUT_ALWAYSCREATEPR: 'false'
    ##[debug] Converted to bool: False
    ##[debug]ENDPOINT_URL_SystemVssConnection: '<removed>'
    ##[debug]ENDPOINT_AUTH_SystemVssConnection: '********'
    ##[debug]ENDPOINT_DATA_SystemVssConnection (empty)
    The repository is: <removed>
    The Source Branch is: refs/heads/release
    The Target Branch is: refs/heads/develop
    The Title is: 🤖 Merge release into develop
    The Description is: Automatically created by "Azure DevOps Create Pull Request Task"
    Is Reviewers are: 
    Is Draft Pull Request: False
    Auto-Complete: True
    Link Work Items: True
```

Replaced some values with `<removed>`.

**Comment by shayki5** [comment]
All good :) 
Is not the full log :/ it stopped in `Link Work Items: True`.
Btw - I just published the new version... but it not will solve your issue...

**Comment by sgabler** [comment]
whoops, I copy&pasted from azure, and that cuts of some data. weird. I now exported the raw log

```
##[debug]Evaluating condition for step: '🤖 Merge release into develop'
##[debug]Evaluating: SucceededNode()
##[debug]Evaluating SucceededNode:
##[debug]=> True
##[debug]Result: True
##[section]Starting: 🤖 Merge release into develop
==============================================================================
Task         : Create Pull Request
Description  : Automatically create a Pull Request for Azure DevOps or GitHub repository from Build or Release pipeline, supports also multi target branch.
Version      : 1.2.277
Author       : Shayki Abramczyk
Help         : [More Information](https://github.com/shayki5/azure-devops-create-pr-task/blob/master/README.md)
==============================================================================
##[debug]VstsTaskSdk 0.10.0 commit <removed>
##[debug]Entering RunTask.
##[debug]INPUT_SOURCEBRANCH: 'release'
##[debug]INPUT_TARGETBRANCH: 'develop'
##[debug]INPUT_TITLE: '🤖 Merge release into develop'
##[debug]INPUT_DESCRIPTION: 'Automatically created by "Azure DevOps Create Pull Request Task"'
##[debug]INPUT_REVIEWERS (empty)
##[debug]INPUT_REPOTYPE: 'Azure DevOps'
##[debug]INPUT_ISDRAFT: 'false'
##[debug] Converted to bool: False
##[debug]INPUT_AUTOCOMPLETE: 'true'
##[debug] Converted to bool: True
##[debug]INPUT_MERGESTRATEGY: 'Merge (No fast-forward)'
##[debug]INPUT_DELETESOURCH (empty)
##[debug] Converted to bool: False
##[debug]INPUT_DELETESOURCE: 'false'
##[debug] Converted to bool: False
##[debug]INPUT_COMMITMESSAGE (empty)
##[debug]INPUT_TRANSITIONWORKITEMS: 'false'
##[debug] Converted to bool: False
##[debug]INPUT_LINKWORKITEMS: 'true'
##[debug] Converted to bool: True
##[debug]INPUT_PROJECTID: <removed>
##[debug]INPUT_GITREPOSITORYID: <removed>
##[debug]INPUT_GITHUBREPOSITORY: <removed>
##[debug]INPUT_PASSPULLREQUESTIDBACKTOADO: 'false'
##[debug] Converted to bool: False
##[debug]INPUT_ISFORKED: 'false'
##[debug] Converted to bool: False
##[debug]INPUT_BYPASSPOLICY: 'false'
##[debug] Converted to bool: False
##[debug]INPUT_BYPASSREASON (empty)
##[debug]INPUT_ALWAYSCREATEPR: 'false'
##[debug] Converted to bool: False
##[debug]ENDPOINT_URL_SystemVssConnection: <removed>
##[debug]ENDPOINT_AUTH_SystemVssConnection: '********'
##[debug]ENDPOINT_DATA_SystemVssConnection (empty)
The repository is: <removed>
The Source Branch is: refs/heads/release
The Target Branch is: refs/heads/develop
The Title is: 🤖 Merge release into develop
The Description is: Automatically created by "Azure DevOps Create Pull Request Task"
Is Reviewers are: 
Is Draft Pull Request: False
Auto-Complete: True
Link Work Items: True
Bypass: False
Bypass Reason: 
DeleteSourceBranch ist set to: False
##[debug]GET <api_url_removed>/_apis/git/repositories/0baf6bfa-ffa8-4b34-9293-cf7462fbd850/diffs/commits?baseVersion=release&targetVersion=develop&api-version=4.0&diffCommonCommit=false&$top=2 with 0-byte payload
##[debug]received 1344-byte response of content type application/json; charset=utf-8; api-version=4.0
1 new commits! perform a Pull Request...
##[debug]POST <api_url_removed>/_apis/git/repositories/0baf6bfa-ffa8-4b34-9293-cf7462fbd850/commitsBatch?api-version=4.0 with -1-byte payload
##[debug]received 739-byte response of content type application/json; charset=utf-8; api-version=4.0
##[debug]@{count=1; value=System.Object[]}
{
    "forkSource":  "",
    "sourceRefName":  "refs/heads/release",
    "title":  "🤖 Merge release into develop",
    "WorkItemRefs":  [
                         null
                     ],
    "description":  "Automatically created by \"Azure DevOps Create Pull Request Task\"",
    "reviewers":  "",
    "isDraft":  "False",
    "targetRefName":  "refs/heads/develop"
}
##[debug]POST <api_url_removed>/_apis/git/repositories/0baf6bfa-ffa8-4b34-9293-cf7462fbd850/pullrequests?api-version=4.0 with -1-byte payload
##[debug]received 5548-byte response of content type application/json; charset=utf-8; api-version=4.0
*************************
******** Success ********
*************************
Pull Request 3222 created.
##[debug]{
##[debug]    "completionOptions":  {
##[debug]                              "mergeStrategy":  "Merge (No fast-forward)",
##[debug]                              "transitionWorkItems":  "False",
##[debug]                              "deleteSourceBranch":  "False",
##[debug]                              "mergeCommitMessage":  ""
##[debug]                          },
##[debug]    "autoCompleteSetBy":  {
##[debug]                              "id":  "af4da24c-b355-40e5-b294-5f84cc37fe28"
##[debug]                          }
##[debug]}
##[debug]<api_url_removed>/_apis/git/repositories/0baf6bfa-ffa8-4b34-9293-cf7462fbd850/pullrequests/3222?api-version=4.1
##[debug]PATCH <api_url_removed>/_apis/git/repositories/0baf6bfa-ffa8-4b34-9293-cf7462fbd850/pullrequests/3222?api-version=4.1 with -1-byte payload
##[warning]Can't set Auto Complete to PR 3222.
##[debug]Processed: ##vso[task.logissue type=warning]Can't set Auto Complete to PR 3222.
##[warning]{"$id":"1","innerException":null,"message":"TF401027: You need the Git 'GenericContribute' permission to perform this action. Details: identity 'Build\\11400df4-d5a0-4a8a-b46c-a4d5632b067a', scope 'branch'.","typeName":"Microsoft.TeamFoundation.Git.Server.GitNeedsPermissionException, Microsoft.TeamFoundation.Git.Server","typeKey":"GitNeedsPermissionException","errorCode":0,"eventId":3000}
##[debug]Processed: ##vso[task.logissue type=warning]{"$id":"1","innerException":null,"message":"TF401027: You need the Git 'GenericContribute' permission to perform this action. Details: identity 'Build\\11400df4-d5a0-4a8a-b46c-a4d5632b067a', scope 'branch'.","typeName":"Microsoft.TeamFoundation.Git.Server.GitNeedsPermissionException, Microsoft.TeamFoundation.Git.Server","typeKey":"GitNeedsPermissionException","errorCode":0,"eventId":3000}
##[warning]The remote server returned an error: (403) Forbidden.
##[debug]Processed: ##vso[task.logissue type=warning]The remote server returned an error: (403) Forbidden.
##[debug]Leaving RunTask.
##[section]Finishing: 🤖 Merge release into develop
```

**Comment by sgabler** [comment]
I'm playing around with the API now. It's returning some unexpected results in some places. For example, when I compare "release" with "develop" branch, there shouldn't be any commits, because "release" is behind the "develop" branch. But the API doesn't show this.

I will let you know what I find out

**Comment by shayki5** [comment]
Thanks for the investigation! 
The branches in the api call are in the right order: `baseVersion=$($sourceBranch)&targetVersion=$($targetBranch)` no?
The only thing is the `diffCommonCommit`, I can switch it to `true` of course, should I?

**Comment by sgabler** [comment]
Let me double check. To do so I just copy&pasted the API call from the task's logs to see how it currently works:

`<api_url>/_apis/git/repositories/0baf6bfa-ffa8-4b34-9293-cf7462fbd850/diffs/commits?baseVersion=release&targetVersion=develop&api-version=4.0&diffCommonCommit=false&$top=2`

aheadCount: 47
behindCount: 1
changeCounts: 126 Edits

Then I change the parameter to true:

`<api_url>/_apis/git/repositories/0baf6bfa-ffa8-4b34-9293-cf7462fbd850/diffs/commits?baseVersion=release&targetVersion=develop&api-version=4.0&diffCommonCommit=true&$top=2`

Nothing changes

Then I swap the branch names:

`<api_url>/_apis/git/repositories/0baf6bfa-ffa8-4b34-9293-cf7462fbd850/diffs/commits?baseVersion=develop&targetVersion=release&api-version=4.0&diffCommonCommit=true&$top=2`

![image](https://user-images.githubusercontent.com/3625477/145247037-bdacff27-625c-46fe-bd57-ef29ddf37ec5.png)

I think the branches need to be swapped, and the parameter needs to be changed.

The response looks correct to me, because I want to merge the release branch into the develop branch, which is 1 commit ahead, 47 commits behind, but there are no actual file changes.

This probably seems confusing, because in #118 the parameter `diffCommonCommit` was first introduced. But I think it only worked accidentally in the scenario that I was testing back then with your help. My branches were a bit differently set up back then.

**Comment by shayki5** [comment]
I think you put the wrong parameters:

```
##[debug]INPUT_SOURCEBRANCH: 'release'
##[debug]INPUT_TARGETBRANCH: 'develop'
```
The `develop` should be the source, no?

So, switch the `diffCommonCommit` to `true`?

**Comment by shayki5** [comment]
When you right you right 😀
So, swap the branches and switch `diffCommonCommit` to `true`?

**Comment by sgabler** [comment]
Exactly 🎉😄

**Comment by shayki5** [comment]
Done and published :)
Can you check please?

**Comment by shayki5** [comment]
@sgabler see issue #127, looks like it's not good, I rollbacked the change.

**Comment by sgabler** [comment]
oh no. I will take a look tomorrow. Sorry for the inconvenience.

**Comment by sgabler** [comment]
Couldn't sleep because I was thinking about this issue 😅. But I think I have a solution.

After swapping the branches, the `behindCount` can't be used anymore and should be replaced with `aheadCount`.

That also actually makes more sense now: we want to know if the source branch is **ahead** of the target branch, not behind. Otherwise a merge can't happen.

So in short, the original changes from commit ed4be96024f0ebab79aeebd2ad892f1d90e8fac6 can be applied again, but additionally the 2 usages of `behindCount` should be replaced with `aheadCount`.

What do you think?

**Comment by shayki5** [comment]
Hi, you're right and I replaced the it to `aheadCount`, but the issue was the `diffCommonCommit=true` that I changed but didn't revert it, when it's `true` the `changeCounts` is empty! and PR not created. I switched it to `false` until we will understand exactly what is the meaning of `true`...

**Comment by sgabler** [comment]
Thank you. I will prepare an explanation and comparison here to explain it. In my scenario the `diffCommonCommit` needs to be `true`, otherwise the feature request from #118 will not work anymore, I believe.

Will post that soon. Thank you for your cooperation so far 🙂

**Comment by sgabler** [comment]
@shayki5 I created a Draft PR: https://github.com/shayki5/azure-devops-create-pr-task/pull/129

I hope this makes what I had in mind a bit clearer.

**Comment by shayki5** [comment]
@sgabler looks good but something it's strange in the API response, I did a test, created a new branch, changes something in a file and I got this (if `diffCommonCommit=true`):

```
{"allChangesIncluded":true,"changeCounts":{},"changes":[],"commonCommit":"73801bc6bcd4763b13a69d888d0b52c721704736","baseCommit":"73801bc6bcd4763b13a69d888d0b52c721704736","targetCommit":"fe3df0a7ea691a14addd0b94a8d130067577dacc","aheadCount":1,"behindCount":0}
```
The `changeCounts` is empty. why? the `aheadCount` is 1 that is correct.
In such case, in your new changes in the draft PR if `alwaysCreatePR=false` PR will not be created 😕

**Comment by sgabler** [comment]
Strange. Here is how I test this. Let me know if there is any difference in the way you test

- I create a new branch from "develop", which is my base branch. This branch is called "test-pr-creation"
- I make one change to 1 file and push it
- Then I check the difference in Azure with the "compare branches feature"
    - There should be 1 commit shown: ✅ 
    ![image](https://user-images.githubusercontent.com/3625477/145380756-912e1f93-b079-44a6-b635-be21cca4e57f.png)
    - There should be file changes in 1 file: ✅ 
    ![image](https://user-images.githubusercontent.com/3625477/145381269-dfeeec8a-936c-42d9-ac8d-c38275c0b40f.png)
- Then I compare this with the API Response in the browser
	- `/diffs/commits?baseVersion=develop&targetVersion=test-pr-creation&api-version=4.0&diffCommonCommit=true&$top=2` should give me 1 aheadCount, 0 behindCount, and it should show some changes: ✅ 
	![image](https://user-images.githubusercontent.com/3625477/145382182-85ea96f6-d432-42c8-a480-42f49f7c4852.png)
	- If i change `diffCommonCommit=false`, I get 100% the same response

Is this also how you test?

Three things:
- Why does it show changeCounts->Edits 3? I think it's because I changed one file (1 change). But this file is located in one directory (1 change) and this directory is in another directory (1 change). In total: 3 changes. I think this is used by Azure to build the directory tree with all changes later
- In this scenario it's expected that the `diffCommonCommit` flag doesn't change the response. I can provide another example later where there would be a difference, but maybe let's just keep it simple in the first step
- I tested the API response in the browser, not via the Powershell script

**Comment by shayki5** [comment]
Hi, I testes it again with new branch and you're right, don't know why before I got something else, maybe I mixed the branches by mistake.
So, switch it to regular PR, I will merge it to master, publish it privately and will test it.

**Comment by sgabler** [comment]
So after #129 was merged, I got access to the private version of this task and tested in my organization.

First test: Merge master into release branch

Expected: 2 commits difference, no PR will be created because no file changes exist. A hint should be shown that the AlwaysCreatePr flag can be enabled.

```
Starting: 🤖 Merge master into release
==============================================================================
Task         : Create Pull Request
Description  : Automatically create a Pull Request for Azure DevOps or GitHub repository from Build or Release pipeline, supports also multi target branch.
Version      : 1.2.283
Author       : Shayki Abramczyk
Help         : [More Information](https://github.com/shayki5/azure-devops-create-pr-task/blob/master/README.md)
==============================================================================
The repository is: <removed>
The Source Branch is: refs/heads/master
The Target Branch is: refs/heads/release
The Title is: 🤖 Merge master into release
The Description is: Automatically created by "Azure DevOps Create Pull Request Task"
Is Reviewers are: 
Is Draft Pull Request: False
Auto-Complete: True
Link Work Items: True
Bypass: False
Bypass Reason: 
DeleteSourceBranch ist set to: False
AlwaysCreatePr flag is false. A PR will only be created if there are actual file changes, but not if there is only a difference in commits.
##[warning]***************************************************************
##[warning]There are no file changes in the source branch, so no PR will be created!
##[warning]The source branch is ahead by 2 commits. If you want to create a PR in such a case, then please set the AlwaysCreatePr flag to true.
##[warning]***************************************************************
Finishing: 🤖 Merge master into release
```

✅ 

--------------------

Second test: Merge release into develop

Expected: also 2 commits difference, no file changes, no PR should be created. A hint should be shown that the AlwaysCreatePr flag can be enabled.

```
Starting: 🤖 Merge release into develop
==============================================================================
Task         : Create Pull Request
Description  : Automatically create a Pull Request for Azure DevOps or GitHub repository from Build or Release pipeline, supports also multi target branch.
Version      : 1.2.283
Author       : Shayki Abramczyk
Help         : [More Information](https://github.com/shayki5/azure-devops-create-pr-task/blob/master/README.md)
==============================================================================
The repository is: <removed>
The Source Branch is: refs/heads/release
The Target Branch is: refs/heads/develop
The Title is: 🤖 Merge release into develop
The Description is: Automatically created by "Azure DevOps Create Pull Request Task"
Is Reviewers are: 
Is Draft Pull Request: False
Auto-Complete: True
Link Work Items: True
Bypass: False
Bypass Reason: 
DeleteSourceBranch ist set to: False
AlwaysCreatePr flag is false. A PR will only be created if there are actual file changes, but not if there is only a difference in commits.
##[warning]***************************************************************
##[warning]There are no file changes in the source branch, so no PR will be created!
##[warning]The source branch is ahead by 2 commits. If you want to create a PR in such a case, then please set the AlwaysCreatePr flag to true.
##[warning]***************************************************************
Finishing: 🤖 Merge release into develop
```

✅ 

------------------------

Third test: Merge develop into release

Expected: Currently the branches are identical, so no changes should happen

```
Starting: 🤖 Merge release into develop
==============================================================================
Task         : Create Pull Request
Description  : Automatically create a Pull Request for Azure DevOps or GitHub repository from Build or Release pipeline, supports also multi target branch.
Version      : 1.2.283
Author       : Shayki Abramczyk
Help         : [More Information](https://github.com/shayki5/azure-devops-create-pr-task/blob/master/README.md)
==============================================================================
The repository is: <removed>
The Source Branch is: refs/heads/develop
The Target Branch is: refs/heads/release
The Title is: 🤖 Merge release into develop
The Description is: Automatically created by "Azure DevOps Create Pull Request Task"
Is Reviewers are: 
Is Draft Pull Request: False
Auto-Complete: True
Link Work Items: True
Bypass: False
Bypass Reason: 
DeleteSourceBranch ist set to: False
AlwaysCreatePr flag is false. A PR will only be created if there are actual file changes, but not if there is only a difference in commits.
##[warning]***************************************************************
##[warning]There are no file changes in the source branch, so no PR will be created!
##[warning]***************************************************************
Finishing: 🤖 Merge release into develop
```

✅ 

--------------

Fourth test: trying the same thing with AlwaysCreatePr set to true

Expected: No pr is created

```
Starting: 🤖 Merge release into develop
==============================================================================
Task         : Create Pull Request
Description  : Automatically create a Pull Request for Azure DevOps or GitHub repository from Build or Release pipeline, supports also multi target branch.
Version      : 1.2.283
Author       : Shayki Abramczyk
Help         : [More Information](https://github.com/shayki5/azure-devops-create-pr-task/blob/master/README.md)
==============================================================================
The repository is: <removed>
The Source Branch is: refs/heads/develop
The Target Branch is: refs/heads/release
The Title is: 🤖 Merge release into develop
The Description is: Automatically created by "Azure DevOps Create Pull Request Task"
Is Reviewers are: 
Is Draft Pull Request: False
Auto-Complete: True
Link Work Items: True
Bypass: False
Bypass Reason: 
DeleteSourceBranch ist set to: False
AlwaysCreatePr flag is true. Trying to perform a Pull Request...
##[warning]***************************************************************
##[warning]There are no new commits in the source branch, no PR is needed!
##[warning]***************************************************************
Finishing: 🤖 Merge release into develop
```

✅ 

---------------

Fifth test: merging a branch into another with 1 commit difference, which also has file changes. AlwaysCreatePR flag is true.

Expected: PR is created

```
View raw log

Starting: 🤖 test
==============================================================================
Task         : Create Pull Request
Description  : Automatically create a Pull Request for Azure DevOps or GitHub repository from Build or Release pipeline, supports also multi target branch.
Version      : 1.2.283
Author       : Shayki Abramczyk
Help         : [More Information](https://github.com/shayki5/azure-devops-create-pr-task/blob/master/README.md)
==============================================================================
The repository is: <removed>
The Source Branch is: refs/heads/azure-pipelines-tests
The Target Branch is: refs/heads/develop
The Title is: 🤖 Merge release into develop
The Description is: Automatically created by "Azure DevOps Create Pull Request Task"
Is Reviewers are: 
Is Draft Pull Request: False
Auto-Complete: True
Link Work Items: True
Bypass: False
Bypass Reason: 
DeleteSourceBranch ist set to: False
AlwaysCreatePr flag is true. Trying to perform a Pull Request...
The source branch is ahead by 3 commits. Perform a Pull Request...
{
    "forkSource":  "",
    "sourceRefName":  "refs/heads/azure-pipelines-tests",
    "title":  "🤖 Merge release into develop",
    "WorkItemRefs":  [
                         null
                     ],
    "description":  "Automatically created by \"Azure DevOps Create Pull Request Task\"",
    "reviewers":  "",
    "isDraft":  "False",
    "targetRefName":  "refs/heads/develop"
}
*************************
******** Success ********
*************************
Pull Request 3229 created.
Finishing: 🤖 test
```

✅ 

----------------

Sixth test: same thing, but with AlwaysCreatePr set to false

Expected: PR is created

```
Starting: 🤖 test
==============================================================================
Task         : Create Pull Request
Description  : Automatically create a Pull Request for Azure DevOps or GitHub repository from Build or Release pipeline, supports also multi target branch.
Version      : 1.2.283
Author       : Shayki Abramczyk
Help         : [More Information](https://github.com/shayki5/azure-devops-create-pr-task/blob/master/README.md)
==============================================================================
The repository is: <removed>
The Source Branch is: refs/heads/azure-pipelines-tests
The Target Branch is: refs/heads/develop
The Title is: 🤖 Merge release into develop
The Description is: Automatically created by "Azure DevOps Create Pull Request Task"
Is Reviewers are: 
Is Draft Pull Request: False
Auto-Complete: True
Link Work Items: True
Bypass: False
Bypass Reason: 
DeleteSourceBranch ist set to: False
AlwaysCreatePr flag is false. A PR will only be created if there are actual file changes, but not if there is only a difference in commits.
File changes were found! Perform a Pull Request...
{
    "forkSource":  "",
    "sourceRefName":  "refs/heads/azure-pipelines-tests",
    "title":  "🤖 Merge release into develop",
    "WorkItemRefs":  [
                         null
                     ],
    "description":  "Automatically created by \"Azure DevOps Create Pull Request Task\"",
    "reviewers":  "",
    "isDraft":  "False",
    "targetRefName":  "refs/heads/develop"
}
*************************
******** Success ********
*************************
Pull Request 3230 created.
Finishing: 🤖 test
```

✅

**Comment by sgabler** [comment]
@shayki5 can you think of anything else to test? if not, i think we are good to release that version and close this issue. Thanks a lot for your help

**Comment by shayki5** [comment]
Looks good! thank you so much for the tests!! 
I just added 1 one small print and updated the public version 😀
